File: plugins/mb-tracker/src/mb_tracker.clj
(ns mb-tracker
  (:import (ij IJ ImagePlus WindowManager ImageListener)
           (ij.measure Measurements)
           (ij.gui ImageCanvas) 
           (javax.swing SwingUtilities, KeyStroke, JComponent, Box, JFrame, JLabel, JPanel,
                        JTabbedPane, JToolBar, JScrollPane, JButton, JList, JTextField,
                        JTextArea, JFileChooser, BoxLayout, BorderFactory, JOptionPane,
                        JSlider)
           (javax.swing.event ChangeListener)
           (java.awt Dimension, BorderLayout, FlowLayout, GridBagLayout, GridBagConstraints, 
                     Insets, Rectangle)
           (java.awt.event ActionListener
                           WindowListener
                           WindowAdapter
                           ComponentListener
                           ComponentAdapter
                           ActionEvent
                           KeyEvent)))

(gen-class
  :name "org.zephyre.MBTracker"
  :implements ["java.lang.Runnable"])

; UI components
(def components (atom {}))
(def init-image (atom nil))

(defn init-image-listener [] "Image listener for displaying the location crosshair."
  (let [image-updated (fn [image] ())]
    (reify ImageListener
      (imageClosed [this image] 
        (IJ/log "Image closed.")
        (.dispose (:main-frame @components)))
      (imageOpened [this image] (IJ/log (format "%s%s" "Image opened: " (.getTitle image))))
      (imageUpdated [this image]
        (if (identical? image @init-image) 
          (IJ/log (format "%s%d" "Image updated: " (.getCurrentSlice image)))))))) 

(def image-listener (atom (init-image-listener)))

(defn register-listener [] "Register listeners for image open/updated/close."
  (ImagePlus/addImageListener @image-listener)
  (IJ/log "ImagePlus/addImageListener"))    

(defn unregister-listener []
  (ImagePlus/removeImageListener @image-listener)
  (IJ/log "ImagePlus/removeImageListener")) 

(defn init-status-bar []
  (let [bar (JToolBar.), label (JLabel. "Idle.")]
    (.add bar label)
    (swap! components assoc :status-label label)
    (identity bar)))

(def atom-thrd (atom nil))
(def atom-image (atom nil))

(defn update-canvas [thrd] "Update the canvas according to new threshold values."
  (let [pixels (.getPixels (.getProcessor @atom-image))]
    (dorun (map (fn [idx]
                  (aset ^shorts pixels idx (short thrd)))
;                  (if (< (aget ^shorts pixels idx) thrd) (aset ^shorts pixels idx (short 0))))
                (range (count thrd))))
    (.updateAndRepaintWindow @atom-image)
    (.repaint (:main-frame @components))))

(defn perform-tracking [image roi thrd])

(defn on-ok-btn [thrd]
  "The OK button is pressed"
  (let [image (WindowManager/getCurrentImage)]
    (if (not image) (IJ/showMessage "Error" "No image is present.")
      (let [roi (try (.getBounds (.getRoi image))
                  (catch NullPointerException e
                    (Rectangle. (.getWidth image) (.getHeight image))))
            data (.getPixels (.getProcessor image))
            width (.getWidth image)
            height (.getHeight image)
            ; [1 2 3 ... 1 2 3 ... ...]
            xseq (apply concat (repeat height (range width)))
            ; [1 1 1 ... 2 2 2 ... 3 3 3 ... ... n n n ...]
            yseq (apply concat (map #(repeat width %) (range height)))]
        (/ (reduce + (map * data xseq)) (reduce + data))))))

(defn init-ctrl-panel [thrd-range]
  (let [panel (JPanel.)
        field (JTextField. (first thrd-range))
        btn (JButton. "Go!")
        image (WindowManager/getCurrentImage)]
;        slider (JSlider. (int (first thrd-range)) (int (last thrd-range)))]
    (doto field
      (.setText "32")
      (.setPreferredSize (Dimension. 32 (.height (.getPreferredSize field)))))
    (.requestFocus btn)
    (.addActionListener btn
      (reify ActionListener
        (actionPerformed [this evt]
          (try 
            (on-ok-btn (Integer/parseInt (.getText field)))
            (catch NumberFormatException e (IJ/showMessage "Error" "Invalid threshold."))))))
    (doto panel
;      (.setBorder panel (BorderFactory/createEmptyBorder 4 4 4 4))
      (.setLayout (FlowLayout.))
      (.add (JLabel. "Threshold:"))
      (.add field)
      (.add btn))))

(defn init-main-panel
  [frame thrd-range]
  (let [panel (JPanel.)
        layout (BorderLayout.)]
    (doto layout
      (.setHgap 8)
      (.setVgap 8))
    (doto panel
      (.setLayout layout)
      (.add (init-ctrl-panel thrd-range) (BorderLayout/NORTH))
      (.add (init-status-bar) (BorderLayout/SOUTH))
      (.setBorder (BorderFactory/createEmptyBorder 4 4 4 4)))
    (let [image (WindowManager/getCurrentImage)]
      (if image        
        (let [img (ImagePlus. "Threshold" (.duplicate (.getProcessor image)))
              canvas (ImageCanvas. img)]
          (.add panel canvas)
          (reset! atom-image img))))
    panel))

(defn init-ui [thrd-range] "Initialize the main frame"
  (let [frame (JFrame. "Magnetic Beads Tracker v0.9 / Zephyre")]
    (.add (.getContentPane frame) (init-main-panel frame thrd-range))
    (.registerKeyboardAction
      (.getRootPane frame)
      (reify ActionListener (actionPerformed [this evt] (.dispose frame)))
      (KeyStroke/getKeyStroke KeyEvent/VK_ESCAPE 0)
      (JComponent/WHEN_IN_FOCUSED_WINDOW))
    (doto frame
      (.setDefaultCloseOperation (JFrame/DISPOSE_ON_CLOSE))
      (.setLocationRelativeTo nil)
      (.addWindowListener
        (proxy [WindowAdapter] []
          (windowOpened [evt] (register-listener))
          (windowClosed [evt] (unregister-listener)))) 
;      (.addComponentListener
;        (proxy [ComponentAdapter] []
;          (componentShown [evt] (register-listener))
;          (componentHidden [evt] (unregister-listener))))
      (.pack))))

(defn min-max-over-stack [image]
  (let [n (.getNSlices image)
        initSlice (.getSlice image)]
    (.setSlice image 1)
    (let [result (reduce (fn [val idx]
                           (.setSlice image idx)
                           (IJ/showProgress idx n)
                           (let [stat (.getStatistics image)]
                             [(min (first val) (.min stat))
                              (max (last val) (.max stat))]))            
                         (let [stat (.getStatistics image)] [(.min stat) (.max stat)])
                         (range 2 (inc n)))]
      (.setSlice image initSlice)
      (identity result))))

(defn -run [this]
  (let [image (WindowManager/getCurrentImage)]
    (if image
      (do
        (let [new-frame (init-ui (min-max-over-stack image))]
          (swap! components assoc :main-frame new-frame)
          (.setVisible new-frame true))
        (reset! init-image image))
      (let [new-frame (init-ui [0 255])]
          (swap! components assoc :main-frame new-frame)
          (.setVisible new-frame true)))))      

File: plugins/mb-tracker/src/MBTracker_.java
import java.awt.Rectangle;
import java.awt.geom.Point2D;
import java.awt.geom.Point2D.Double;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.URLClassLoader;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executors;

import ij.IJ;
import ij.ImagePlus;
import ij.WindowManager;
import ij.gui.GenericDialog;
import ij.gui.Roi;
import ij.measure.Measurements;
import ij.measure.ResultsTable;
import ij.plugin.PlugIn;
import ij.process.ImageStatistics;

public class MBTracker_ implements PlugIn {

//	public void run2(String arg0) {
//		final URLClassLoader loader = (URLClassLoader) getClass()
//				.getClassLoader();
//		try {
//			Class<?> cls = (Executors.newFixedThreadPool(1)
//					.submit(new Callable<Class<?>>() {
//						@Override
//						public Class<?> call() throws Exception {
//							Thread.currentThread()
//									.setContextClassLoader(loader);
//							return Class.forName("org.zephyre.MBTracker");
//						}
//					})).get();
//			((Runnable) (cls.newInstance())).run();
//		} catch (InterruptedException | ExecutionException
//				| InstantiationException | IllegalAccessException
//				| SecurityException | IllegalArgumentException e) {
//			IJ.log(e.toString());
//		}
//	}

	private static MBTrackerListener listener_;
	static {
		listener_ = new MBTrackerListener();
		ImagePlus.addImageListener(listener_);
	}

	@Override
	public void run(String arg0) {
		ImagePlus image = WindowManager.getCurrentImage();
		if (image == null) {
			IJ.showMessage("Error", "No image opened.");
			return;
		}

		GenericDialog gd = new GenericDialog("Magnetic Beads Tracker");
		gd.addMessage("Powered by Haizi Zheng, 2013");
		gd.addMessage("");
		gd.addNumericField("Threshold: ", 0, 0);
		gd.showDialog();
		if (gd.wasCanceled())
			return;
		int threshold = (int) gd.getNextNumber();

		Random rand = new Random();
		List<Point2D> track = new ArrayList<Point2D>();

		Roi roi = image.getRoi();
		Rectangle roiRc;
		if (roi != null)
			roiRc = roi.getBounds();
		else
			roiRc = new Rectangle(image.getWidth(), image.getHeight());
		int rcW = roiRc.width;
		int rcH = roiRc.height;

		for (int i = 0; i < image.getNSlices(); i++) {
			image.setSlice(i + 1);
			Point2D trackPt = performTracking(image, roiRc, threshold);
			track.add(trackPt);

			int x = (int) (trackPt.getX() - 0.5 * rcW);
			int y = (int) (trackPt.getY() - 0.5 * rcH);
			if (x < 0)
				x = 0;
			else if (x + rcW >= image.getWidth())
				x = image.getWidth() - rcW;
			if (y < 0)
				y = 0;
			else if (y + rcH >= image.getHeight())
				y = image.getHeight() - rcH;
			roiRc = new Rectangle(x, y, rcW, rcH);
		}

		renderResultsTable(track);
		listener_.addTrackingResult(image, track);
		listener_.imageUpdated(image);
		image.updateAndDraw();
		
		// ImagePlus.addImageListener(listener_);
	}

	private void renderResultsTable(List<Point2D> track) {
		ResultsTable table = new ResultsTable();
		for (Point2D pt:track){
			table.incrementCounter();
			table.addValue("X", pt.getX());
			table.addValue("Y", pt.getY());
		}
		table.show("Results");
	}

	private static Point2D performTracking(ImagePlus image, Rectangle roiRc,
			int threshold) {
		short[] data = (short[]) image.getProcessor().getPixels();
		int width = image.getWidth();

		double sumx = 0, sumy = 0, sum = 0;
		for (int i = roiRc.y; i < roiRc.y + roiRc.height; i++) {
			for (int j = roiRc.x; j < roiRc.x + roiRc.height; j++) {
				short val = data[i * width + j];
				if (val < threshold)
					val = 0;
				sumx += val * j;
				sumy += val * i;
				sum += val;
			}
		}
		return new Point2D.Double(sumx / sum, sumy / sum);
	}
}
